pkg/cmd/pulumi/packagecmd/package_publish.go

```diff
@@ -15,162 +15,232 @@
+	} else { // As a last resort, try to determine the publisher from the default organization or fail if none is found.
+		ws := pkgWorkspace.Instance
+		project, _, err := ws.ReadProject()
+		if err != nil && !errors.Is(err, workspace.ErrProjectNotFound) {
+			return fmt.Errorf("failed to determine default organization: %w", err)
+		}
+		publisher, err = cmd.defaultOrg(project)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to determine default organization: %w", err)
+		}
+		if publisher == "" {
+			return errors.New("no publisher specified and no default organization found, please set a publisher in " +
+				"the package schema or set a default organization in your pulumi config")
 		}
-	case "all", "python", "java", "dotnet":
-		return fmt.Errorf("support for %q coming soon", lang)
-
-	default:
-		return fmt.Errorf("unsupported language %q", lang)
 	}
 
-	return nil
-}
+	name := pkg.Name
+	if name == "" {
+		return errors.New("no package name specified, please set one in the package schema")
+	}
+	var version semver.Version
+	if pkg.Version != nil {
+		version = *pkg.Version
+	} else {
+		return errors.New("no version specified, please set a version in the package schema")
+	}
 
-func publishToNPM(path string) error {
-	// verify path
-	info, err := os.Stat(path)
+	json, err := pkg.MarshalJSON()
 	if err != nil {
-		return fmt.Errorf("reading path %s: %w", path, err)
+		return fmt.Errorf("failed to marshal schema: %w", err)
 	}
-	if !info.IsDir() {
-		return fmt.Errorf("not a directory: %s", path)
+
+	registry, err := b.GetPackageRegistry()
+	if err != nil {
+		return fmt.Errorf("failed to get package registry: %w", err)
 	}
 
-	// Verify npm exists and is set up: npm, user login
-	npm, err := executable.FindExecutable("npm")
+	// In the future we could slurp up the readme from the package source and add the necessary markdown headers.
+	readme, err := os.Open(args.readmePath)
 	if err != nil {
-		return fmt.Errorf("npm whoami: %w", err)
+		return fmt.Errorf("failed to open readme file: %w", err)
 	}
+	defer contract.IgnoreClose(readme)
 
-	// verify auth for npm
-	whoamiCmd := exec.Command(npm, "whoami")
-	whoamiCmd.Stderr = os.Stderr
-	whoami, err := whoamiCmd.Output()
+	var installDocs *os.File
+	if args.installDocsPath != "" {
+		installDocs, err = os.Open(args.installDocsPath)
+		if err != nil {
+			return fmt.Errorf("failed to open install docs file: %w", err)
+		}
+		defer contract.IgnoreClose(installDocs)
+	}
+
+	err = registry.Publish(ctx, backend.PackagePublishOp{
+		Source:      args.source,
+		Publisher:   publisher,
+		Name:        name,
+		Version:     version,
+		Schema:      bytes.NewReader(json),
+		Readme:      readme,
+		InstallDocs: installDocs,
+	})
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to publish package: %w", err)
 	}
 
-	logging.V(1).Infof("Logged in as %s", whoami)
+	fmt.Printf("Successfully published package %s/%s@%s\n", publisher, name, version.String())
```

Comment:
Nit: You don't need to call `.String()` when you format with `%s`.
```suggestion
	fmt.Printf("Successfully published package %s/%s@%s\n", publisher, name, version)
```